google/cloud/aiplatform/lro.py

```diff
@@ -15,6 +15,90 @@
 # limitations under the License.
 #
 
+import proto
+
+from google.api_core import operation as ga_operation
+from google.cloud.aiplatform import base
+from typing import Optional, Callable
+
 
 class LRO:
-    pass
+    """A handler for operation futures."""
+
+    def __init__(
+        self,
+        operation_future: ga_operation.Operation,
+        resource_noun_obj: Optional[base.AiPlatformResourceNoun] = None,
+        result_key: Optional[str] = None,
+        api_get: Optional[Callable[..., proto.Message]] = None,
+    ):
+        """Initialises class with operation and optional object to update.
+
+        Args:
+            operation_future (ga_operation.Operation):
+                Required. Operation future to handle.
+            resource_noun_obj (base.AiPlatformResourceNoun):
+                Optional. Resource to be updated upon operation completion.
+                result_key and api_get also need to be passed.
+            result_key (str):
+                Optional. Attribute to retrieve from result.
+                resource_noun_obj and api_get also need to be passed.
+            api_get (Callable[...,proto.Message]):
+                Optional. Callable that takes resource name and returns object.
+                resource_noun_obj and result_key also need to be passed.
+        """
+        self._operation_future = operation_future
+        if all([resource_noun_obj, result_key, api_get]):
```

Comment:
Should an exception be thrown if api_get if passed without `resource_noun_obj` and `result_key` being passed as well.

Otherwise, could lead to some hard to debug situations if the user misses that requirement.

Reply:
I'm trying to think of a more explicit way to show required arguments.

In Java, multiple constructors could solve this. Swift could solve this with multiple constructors or associated value enums.

If there's a more Pythonic way to express this (rather than just hoping the user reads the documentation), I'd love to know.

Reply:
+1 for raising if an unsupported combination is passed. 

Here's an example in the generated client constructor:
https://github.com/googleapis/python-automl/blob/0233804885846295508a5fc98929dba598172244/google/cloud/automl_v1/services/prediction_service/client.py#L233-L248

Reply:
One way would be to use a class method:

```
    def __init__(
        self,
        operation_future: ga_operation.Operation):
        self._operation_future = operation_future

    @classmethod
    def make_with_callback(
        self,
        operation_future: ga_operation.Operation,
        resource_noun_obj: base.AiPlatformResourceNoun,
        result_key: str,
        api_get: Callable[..., proto.Message],
    ):
        obj = cls.__new__(cls)  # Does not call __init__
        super(LRO, obj).__init__(operation_future)
        if all([resource_noun_obj, result_key, api_get]):
            self.add_update_resource_callback(resource_noun_obj, result_key, api_get)
        else:
            # throw exception
```

This is my attempt at wrapping my head around Python so I'm open to comments.

Reply:
Depends on what you mean by show.

https://docs.python.org/3/library/typing.html#typing.overload
You still have to do the same song and dance in a single method and it's up to the introspection to respect the overloaded signatures.

Reply:
I've come up with a quick solution for this for now. if we think users will use LRO directly, we can improve it in a future PR.